src/multirust-cli/self_update.rs

```diff
@@ -0,0 +1,1000 @@
+//! Self-installation and updating
+//!
+//! This is the installer at the heart of Rust. If it breaks
+//! everything breaks. It is conceptually very simple, as multirust is
+//! distributed as a single binary, and installation mostly requires
+//! copying it into place. There are some tricky bits though, mostly
+//! because of workarounds to self-delete an exe on Windows.
+//!
+//! During install (as `multirust-setup`):
+//!
+//! * copy the self exe to $CARGO_HOME/bin
+//! * hardlink rustc, etc to *that*
+//! * update the PATH in a system-specific way
+//! * run the equivalent of `multirust default stable`
+//!
+//! During upgrade (`multirust self upgrade`):
+//!
+//! * download multirust-setup to $CARGO_HOME/bin/multirust-setup
+//! * run multirust-setup with appropriate flags to indicate
+//!   this is a self-upgrade
+//! * multirust-setup copies bins and handlinks into place. On windows
+//!   this happens *after* the upgrade command exits successfully.
```

Comment:
multirust-setup can't replace the running exe until it exits on Windows. In my experiments I also couldn't mess with the hardlinks until the running exe exits.